File: Stack/4. Generate Parentheses/solution.exs
defmodule Solution do
  @spec generate_parenthesis(n :: integer) :: [String.t()]
  def generate_parenthesis(n) do
    recursive(0, 0, "", n, [])
  end

  def recursive(left, right, curr, n, result) when left == n and right == n, do: [curr | result]

  def recursive(left, right, curr, n, result) do
    l = (left < n && recursive(left + 1, right, "#{curr}(", n, result)) || result
    (right < left && recursive(left, right + 1, "#{curr})", n, l)) || l
  end
end

File: Stack/4. Generate Parentheses/solution.rb
# @param {Integer} n
# @return {String[]}
def generate_parenthesis(n)
  recursive(0, 0, n, "", [])
end

def recursive(left, right, n, curr, result)
  return result.push(curr) if left == n and right == n

  l = left < n ? recursive(left + 1, right, n, "#{curr}(", result) : result
  right < left ? recursive(left, right + 1, n, "#{curr})", l) : l
end

puts "#{generate_parenthesis(3)}"
